lib/LanguagePicker.js

```diff
@@ -19,50 +42,83 @@ function LanguagePicker(lang = 'en', config = {}) {
   // Use the given language as first choice
   fallbacks = [lang].concat(fallbacks);
 
-  // Fallback to English
-  if (fallbacks.indexOf('en') === -1) {
-    fallbacks.push('en');
-  }
-
-  // Fallback to nameTag
-  if (this.nameTag) {
-    fallbacks.push(this.nameTag);
-  }
+  // Remove duplicates
+  fallbacks = fallbacks.filter((item, i) => fallbacks.indexOf(item) === i);
 
   // Add prefix to all languages if exists
   // eslint-disable-next-line arrow-body-style
   fallbacks = fallbacks.map((code) => {
     return code === this.nameTag ? code : this.prefix + code;
   });
 
+  // Store initial fallbacks
   this.fallbacks = fallbacks;
 }
 
 LanguagePicker.prototype.newProcessor = function newProcessor() {
-  const { fallbacks, nameTag, prefix } = this;
-  let bestChoiceLang = nameTag || (prefix ? `${prefix}en` : 'en');
-  let bestChoiceValue;
   let firstChoiceValue;
 
   return {
-    addValue(lang, value) {
-      const langIndex = fallbacks.indexOf(lang);
-      if (
-        langIndex > -1 &&
-        langIndex <= fallbacks.indexOf(bestChoiceLang)
-      ) {
-        // Given language is a better fallback
-        bestChoiceLang = lang;
-        bestChoiceValue = value;
+    addValue: (lang, value) => {
+      this.values[lang] = value;
+
+      if (!firstChoiceValue) {
+        firstChoiceValue = value;
+      }
+    },
+    getResult: () => {
+      const valueLangCodes = Object.keys(this.values);
+
+      // Get the best value from the best language fallback:
+      // 1. Requested language
+      if (this.values[this.userLang]) {
+        return this.values[this.userLang];
+      }
+
+      // 2. Fallback language from fallbacks.json
+      for (let i = 0; i < this.fallbacks.length; i++) {
+        if (this.values[this.fallbacks[i]]) {
+          return this.values[this.fallbacks[i]];
+        }
+      }
+
+      // 3. Any language with suffix of same script as requested language
+      for (let i = 0; i < valueLangCodes.length; i++) {
+        if (valueLangCodes[i].endsWith(this.langScript)) {
+          return this.values[valueLangCodes[i]];
+        }
+      }
+
+      // 4. Another language with the same script
+      for (let i = 0; i < this.languagesInScript.length; i++) {
+        if (this.values[this.languagesInScript[i]]) {
+          return this.values[this.languagesInScript[i]];
+        }
+      }
+
+      // 5. English
+      if (this.values[this.prefix ? `${this.prefix}en` : 'en']) {
+        return this.values[this.prefix ? `${this.prefix}en` : 'en'];
+      }
+
+      // 6. Any language with suffix 'Latn'
+      // Only do this if the script isn't already latin;
+      // otherwise, we've done it in step 3
+      for (let i = 0; i < valueLangCodes.length; i++) {
+        if (valueLangCodes[i].endsWith('Latn')) {
+          return this.values[valueLangCodes[i]];
+        }
       }
 
```

Comment:
you need two more cases - ends with `'_rm'` https://taginfo.openstreetmap.org/search?q=_rm and `'_pinyin'`  https://taginfo.openstreetmap.org/search?q=pinyin  (they are good if name:en is missing)

Reply:
Yes, we were discussing this -- I want to add it as a separate commit on top; my plan is this (see if you like it?) --

Add a script_aliases.json file, where we define something like this -

{
   "Latn": "ro", "rm"
}
etc

Then, have the code add a fallback that builds a proper lookup (with regexp to search the end with - and _ variances) for the aliases of language scripts. This way, we could relatively easily add aliases we've missed or need to adjust, and we won't have to rely on what *seems* like a standard in OSM and may change (I'm worried about those nonstandard standards...)

But since this whole thing is already a little convoluted (look at the tests! ;) I figured I will add this as a subsequent commit, and we can review that separately and make sure that works right.

What do you think?

Reply:
ja_rm does not mean a language code, it means that this string is a romanized version of the Japanese string (transliteration).  See [osm docs](https://wiki.openstreetmap.org/wiki/Multilingual_names#Japan).  It seems that some regions just chose to use `-Latn`, and some tend to use `_rm`.  Historical, but that means we should treat them similarly.  Lookup table converting ro to Latn wouldn't help.  That's why I think you should simply go through each case one by one (just like you did), and since we now pretty much know them all (less than 10 of them total), we might as well spell them out.  No need to do automagic of "-" vs "_" -- they don't mix.